File: internal/scanner.go
package internal

import (
	"fmt"
	"strconv"

	"github.com/aland20/iad/internal/utils"
)

type Scanner struct {
	Source  string // The consumed line
	Tokens  []Token
	Start   uint16 // Start of each token
	Current uint16 // Current char position in the line
	Line    uint16
}

func NewScanner(source string) *Scanner {

	return &Scanner{
		Source: source,
	}
}

func (s *Scanner) ScanTokens() []Token {

	for !s.isAtEnd() {
		s.Start = s.Current
		s.ScanToken()
	}

	s.Tokens = append(s.Tokens, Token{
		TokenType: EOF,
		Word:      "",
		Value:     "",
		Line:      s.Line,
	})

	return s.Tokens
}

func (s *Scanner) ScanToken() {

	c := s.peek()

	switch c {
	// Skip unnecessary lines
	case ' ', '\r', '\t':
		break
	case '\n':
		s.advanceLine()
		// Basic operations & chars
	case '(':
		s.addToken(LEFT_PAREN, "")
	case ')':
		s.addToken(RIGHT_PAREN, "")
	case '{':
		s.addToken(LEFT_BRACE, "")
	case '}':
		s.addToken(RIGHT_BRACE, "")
	case ',':
		s.addToken(COMMA, "")
	case '.':
		s.addToken(DOT, "")
	case ';':
		s.addToken(SEMICOLON, "")
	case '+':
		s.addToken(PLUS, "")
	case '-':
		s.addToken(MINUS, "")
	case '*':
		s.addToken(STAR, "")
		// Multiple Characters
	case '!':
		token := BANG
		if s.match('=') {
			token = BANG_EQUAL
		}
		s.addToken(token, "")
	case '=':
		token := EQUAL
		if s.match('=') {
			token = EQUAL_EQUAL
		}
		s.addToken(token, "")
	case '>':
		token := GREATER
		if s.match('=') {
			token = GREATER_EQUAL
		}
		s.addToken(token, "")
	case '<':
		token := LESS
		if s.match('=') {
			token = LESS_EQUAL
		}
		s.addToken(token, "")
		// Slash and Comment
	case '/':
		if s.match('/') {
			for s.peek() != '\n' && !s.isAtEnd() {
				s.advance()
			}
		} else {
			s.addToken(SLASH, "")
		}
		// String literals
	case '"':
		s.string()
	default:
		// Number Literal
		if utils.IsDigit(c) {
			s.number()
		} else if utils.IsAlpha(c) {
			s.identifier()
		} else {
			// utils.Error(s.getLine(), "Unexpected character")
			utils.Report(s.getLine(), s.getCurrent(), "Unexpected character")
		}
	}

	// Advance to next character
	s.advance()
}

func (s *Scanner) isAtEnd() bool {
	return s.Current >= uint16(len(s.Source))
}

func (s *Scanner) advance() {
	s.Current = s.Current + 1
}

func (s *Scanner) advanceLine() {
	s.Line = s.Line + 1
}

func (s *Scanner) getLine() string {
	return fmt.Sprint(s.Line)
}

func (s *Scanner) getCurrent() string {
	return fmt.Sprint(s.Current)
}

func (s *Scanner) addToken(tokenType TokenType, value any) {

	end := s.Current
	if s.Start == s.Current {
		end = s.Current + 1
	}

	word := s.Source[s.Start:end]

	s.Tokens = append(s.Tokens, Token{
		TokenType: tokenType,
		Word:      word,
		Value:     value,
		Line:      s.Line,
	})
}

func (s *Scanner) match(expected byte) bool {

	if s.isAtEnd() {
		return false
	}

	if s.peekNext() != expected {
		return false
	}

	s.Current = s.Current + 1
	return true
}

func (s *Scanner) peek() byte {
	if s.isAtEnd() {
		return 0
	}

	return s.Source[s.Current]
}

func (s *Scanner) peekNext() byte {
	if s.Current+1 >= uint16(len(s.Source)) {
		return 0
	}

	return s.Source[s.Current+1]
}

func (s *Scanner) string() {
	for s.peekNext() != '"' && !s.isAtEnd() {
		// Supports multi-line string values
		// We have to advance when we hit new line
		if s.peek() == '\n' {
			s.advanceLine()
		}
		s.advance()
	}

	// Must be closed in one line
	if s.isAtEnd() {
		utils.Error(s.getLine(), "Missing string quote")
	}

	// Closing quote
	s.advance()
	// Trim surrounding quotes
	s.Start = s.Start + 1
	value := s.Source[s.Start:s.Current]
	// Add Token with value
	s.addToken(STRING, value)
}

func (s *Scanner) number() {
	for utils.IsDigit(s.peek()) {
		s.advance()
	}

	// Look for fractional part
	if s.peek() == '.' && utils.IsDigit(s.peekNext()) {
		// Consume the "."
		s.advance()

		for utils.IsDigit(s.peek()) {
			s.advance()
		}
	}
	value := s.Source[s.Start:s.Current]
	realValue, err := strconv.ParseFloat(value, 64)

	if err != nil {
		utils.Error(s.getLine(), "invalid number")
	}

	s.addToken(NUMBER, realValue)
}

func (s *Scanner) identifier() {
	for utils.IsAlphaNumeric(s.peek()) {
		s.advance()
	}

	text := s.Source[s.Start:s.Current]
	token := GetIdentifier(text)
	s.addToken(token, "")
}

File: src/scanner.rs
use crate::{
    error::Error,
    error::ErrorKind,
    token::{self, Line, Number, Token, TokenKind, TokenValue},
};

pub type Index = usize;

#[derive(Debug)]
pub struct Scanner<'a> {
    source: &'a [u8], // The consumed line
    start: Index,     // Start of each token
    current: Index,   // Current char position in the line
    line: Line,
}

impl<'a> Scanner<'a> {
    pub fn new(source: &'a [u8]) -> Self {
        Scanner {
            source,
            start: 0,
            current: 0,
            line: 0,
        }
    }

    pub fn scan_tokens(&mut self) -> Result<Vec<Token>, Vec<Error>> {
        let mut tokens = vec![];
        let mut errors = vec![];

        while !self.is_at_end() {
            self.start = self.current;
            match self.scan_token() {
                Ok(Some(token)) => tokens.push(token),
                Ok(None) => {}
                Err(error) => errors.push(error),
            }
        }

        tokens.push(token::Token::new(
            token::TokenKind::Eof,
            String::new(),
            token::TokenValue::String(String::new()),
            self.line,
        ));

        if errors.is_empty() {
            Ok(tokens)
        } else {
            Err(errors)
        }
    }

    fn scan_token(&mut self) -> Result<Option<Token>, Error> {
        let c = self.advance();

        let empty_value = TokenValue::String(String::new());

        match c {
            // Skip unnecessary lines
            b' ' | b'\r' | b'\t' => Ok(None),
            b'\n' => {
                self.advance_line();
                Ok(None)
            }
            // Basic operations & chars
            b'(' => Ok(Some(self.make_token(TokenKind::LeftParen, empty_value))),
            b')' => Ok(Some(self.make_token(TokenKind::RightParen, empty_value))),
            b'{' => Ok(Some(self.make_token(TokenKind::LeftBrace, empty_value))),
            b'}' => Ok(Some(self.make_token(TokenKind::RightBrace, empty_value))),
            b',' => Ok(Some(self.make_token(TokenKind::Comma, empty_value))),
            b'.' => Ok(Some(self.make_token(TokenKind::Dot, empty_value))),
            b';' => Ok(Some(self.make_token(TokenKind::Semicolon, empty_value))),
            b'+' => Ok(Some(self.make_token(TokenKind::Plus, empty_value))),
            b'-' => Ok(Some(self.make_token(TokenKind::Minus, empty_value))),
            b'*' => Ok(Some(self.make_token(TokenKind::Star, empty_value))),
            // Multiple Characters
            b'!' => {
                let mut token = TokenKind::Bang;
                if self.matches(b'=') {
                    token = TokenKind::BangEqual;
                }
                Ok(Some(self.make_token(token, empty_value)))
            }
            b'=' => {
                let mut token = TokenKind::Equal;
                if self.matches(b'=') {
                    token = TokenKind::EqualEqual;
                }
                Ok(Some(self.make_token(token, empty_value)))
            }
            b'>' => {
                let mut token = TokenKind::Greater;
                if self.matches(b'=') {
                    token = TokenKind::GreaterEqual;
                }
                Ok(Some(self.make_token(token, empty_value)))
            }
            b'<' => {
                let mut token = TokenKind::Less;
                if self.matches(b'=') {
                    token = TokenKind::LessEqual;
                }
                Ok(Some(self.make_token(token, empty_value)))
            }
            // Slash and Comments
            b'/' => {
                if self.matches(b'/') {
                    while self.peek() != b'\n' && !self.is_at_end() {
                        self.advance();
                    }
                    Ok(None)
                } else {
                    Ok(Some(self.make_token(TokenKind::Slash, empty_value)))
                }
            }
            // String Literals
            b'"' => Ok(Some(self.string())),
            _ => {
                // Number Literal
                if c.is_ascii_digit() {
                    Ok(Some(self.number()))
                } else if c.is_ascii_alphanumeric() {
                    Ok(Some(self.identifier()))
                } else {
                    Err(Error {
                        kind: ErrorKind::UnexpectedCharacter,
                        line: self.line,
                        description: String::from(format!("Unexpected '{}' Character", c)),
                        start: self.start,
                        end: self.current,
                    })
                }
            }
        }
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> u8 {
        let c = self.source[self.current];
        self.current += 1;
        c
    }

    fn advance_line(&mut self) {
        self.line += 1;
    }

    fn peek_offset(&self, offset: Index) -> u8 {
        *self.source.get(self.current + offset).unwrap_or(&b'\0')
    }

    fn peek(&self) -> u8 {
        self.peek_offset(0)
    }

    fn peek_next(&self) -> u8 {
        self.peek_offset(1)
    }

    fn make_token(&mut self, kind: TokenKind, value: TokenValue) -> Token {
        let word = self.substring(self.start, self.current);

        Token::new(kind, word, value, self.line)
    }

    fn substring(&self, start: Index, end: Index) -> String {
        String::from_utf8(self.source[start..end].to_vec()).unwrap_or_else(|source| {
            println!("Error: {}", start + source.utf8_error().valid_up_to());
            String::new()
        })
    }

    fn matches(&mut self, c: u8) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.peek() != c {
            return false;
        }

        self.current += 1;
        true
    }

    fn string(&mut self) -> Token {
        while self.peek() != b'"' && !self.is_at_end() {
            // Supports multi-line string values
            // We have to advance when we hit new line
            if self.peek() == b'\n' {
                self.advance_line();
            }

            self.advance();
        }

        // Must be closed in one line
        if self.is_at_end() {
            println!("Line {}: Missing closing quotation", self.line);
        }

        // Closing quote
        self.advance();
        let value = self.substring(self.start + 1, self.current - 1);
        // Add token with value
        self.make_token(TokenKind::String, TokenValue::String(value))
    }

    fn number(&mut self) -> Token {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        // Look for fractional part
        if self.peek() == b'.' && self.peek_next().is_ascii_digit() {
            //consume the "."
            self.advance();

            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        let value: f64 = self
            .substring(self.start, self.current)
            .parse::<Number>()
            .map_err(|_| println!("Line {}: Invalid number", self.line))
            .unwrap();

        self.make_token(TokenKind::Number, TokenValue::Number(value))
    }

    fn identifier(&mut self) -> Token {
        while self.peek().is_ascii_alphanumeric() {
            self.advance();
        }

        let text = self.substring(self.start, self.current);

        self.make_token(TokenKind::Identifier, TokenValue::String(text))
    }
}
